### PR TITLE
Milestones Implementation

### PR DESCRIPTION
**Pull request by anshalshukla**

### CODE REVIEW

File: common/gererics/empty.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package gererics
```
Comment (manav2401) [inline common/gererics/empty.go:1]: Generics 😅

File: consensus/bor/heimdallgrpc/milestone.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+	return res.Result.Result, nil
+}
+
+func (h *HeimdallGRPCClient) FetchNoAckMilestone(ctx context.Context, milestoneID string) error {
+	req := &proto.FetchMilestoneNoAckRequest{
+		MilestoneID: milestoneID,
+	}
+
+	log.Info("Fetching no ack milestone", "milestoneaID", milestoneID)
```
Comment (manav2401) [inline consensus/bor/heimdallgrpc/milestone.go:68]: Minor typo here and at line 79 as well

File: eth/borfinality/whitelist/finality.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		return false, nil
+	}
+
+	res, err := isValidChain(currentHeader, chain, f.doExist, f.Number, f.Hash, f.interval)
+
+	return res, err
+}
+
+func (f *finality[T]) Process(block uint64, hash common.Hash) {
+	f.doExist = true
```
Comment (manav2401) [inline eth/borfinality/whitelist/finality.go:61]: Won't we need a lock here?

File: eth/borfinality/whitelist/milestone.go [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+	// TODO: Uncomment once metrics is added
+	// whitelistedMilestoneMeter.Update(int64(block))
+
+	m.UnlockSprint(block)
+}
+
+// This function will Lock the mutex at the time of voting
+// fixme: get rid of it
+func (m *milestone) LockMutex(endBlockNum uint64) bool {
+	m.finality.Lock()
```
Comment (manav2401) [inline eth/borfinality/whitelist/milestone.go:138]: Can you confirm once if m.finality.Unlock() is not needed?

File: eth/borfinality/bor_checkpoint_verifier.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
+// nolint
+package borfinality
```
Comment (manav2401) [inline eth/borfinality/bor_checkpoint_verifier.go:2]: If possible, let's rename this file to a generic name like bor_verifier.go

File: eth/borfinality/handler_bor.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	// fixme: handle different types of errors
+	if errors.Is(err, ErrNotInRejectedList) {
+		log.Warn("MilestoneID not in rejected list", "milestoneID", milestoneID, "err", err)
+		return err
+	}
+
+	if err != nil {
+		log.Error("Failed to fetch no-ack milestone by ID ", "milestoneID", milestoneID, "err", err)
+
+		return errMilestone
```
Comment (manav2401) [inline eth/borfinality/handler_bor.go:119]: Please create a new error type for this fetching no ack by ID. `errMilestone` is only used for fetching new milestone.

File: eth/borfinality/whitelist/finality.go (hunk 3 above)
Comment (manav2401) [reply eth/borfinality/whitelist/finality.go:61]: Okay I think we only have a RLock above so won't need in Process while writing but we'll still confirm.

File: turbo/app/import.go [hunk 7]
```diff
@@ -205,6 +207,18 @@ func missingBlocks(chainDB kv.RwDB, blocks []*types.Block) []*types.Block {
...
+	tx, err := ethereum.ChainDB().BeginRo(context.Background())
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	// Check the validity of incoming chain - only in case of bor consensus
+	currentHead := rawdb.ReadCurrentHeader(tx)
+	if isValid, _ := borfinality.ValidateReorg(currentHead, chain.Headers, ethereum.ChainConfig()); !isValid {
+		return whitelist.ErrMismatch
```
Comment (manav2401) [inline turbo/app/import.go:219]: We should think on improving error handling here. Also, the errors populated in the child functions are not populated here well.

File: eth/borfinality/whitelist/milestone.go (hunk 4 above)
Comment (anshalshukla) [reply eth/borfinality/whitelist/milestone.go:138]: We have a separate function `UnlockMutex`

File: consensus/bor/heimdallgrpc/milestone.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+	"github.com/ledgerwatch/erigon/consensus/bor/heimdall/milestone"
+	"github.com/ledgerwatch/log/v3"
+
+	proto "github.com/maticnetwork/polyproto/heimdall"
+	protoutils "github.com/maticnetwork/polyproto/utils"
+)
+
+func (h *HeimdallGRPCClient) FetchMilestoneCount(ctx context.Context) (int64, error) {
+	log.Info("Fetching milestone count")
```
[the remark this answers is not in the document]
Comment (anshalshukla) [reply consensus/bor/heimdallgrpc/milestone.go:16]: We have them as Info logs in bor as well.

**Comment by battlmonstr** [comment]
superseded by https://github.com/ledgerwatch/erigon/pull/8091

**Pull request closed without merging**